File: web/src/lib/billing.js
/**
 * Flat-tariff billing arithmetic.
 *
 * This module has no imports on purpose: it is a leaf so that both the app
 * bundle and the dependency-free `scripts/check-billing.mjs` verifier can
 * load it.
 *
 * It is a deliberate mirror of `WattWise-AI/src/billing.py`. Both sides are
 * held to the same expected values by `billingCases.json`, which the Python
 * suite asserts against and the Node script asserts here, so the two
 * implementations cannot drift apart.
 *
 * The contract:
 *
 *   consumptionKwh = dailyKwh * days
 *   energyCharge   = consumptionKwh * tariffPerKwh
 *   estimatedBill  = energyCharge + fixedChargePerPeriod
 *
 * `fixedChargePerPeriod` is added once for the whole period, not once per
 * day. There is no time-of-use banding, no slab structure, no tax, and no
 * proration of a standing charge levied daily.
 */

export const DEFAULT_BILLING_DAYS = 30
export const MAX_BILLING_DAYS = 365

export const BILLING_FORMULA =
  'consumption_kwh = daily_kwh * days; ' +
  'energy_charge = consumption_kwh * tariff_per_kwh; ' +
  'estimated_bill = energy_charge + fixed_charge_per_period'

/**
 * Round half away from zero, matching Python's `Decimal(ROUND_HALF_UP)`.
 *
 * Not `Number.prototype.toFixed`: that returns a string, and the two do not
 * always agree on values with more than 15 significant digits. Not plain
 * `Math.round` either, because `1.005 * 100` is `100.49999999999999` in
 * binary floating point and would round down to `1` when the decimal intent
 * was `1.01`, so a small epsilon nudge is applied first.
 */
export function roundHalfUp(value, places = 2) {
  const numeric = Number(value)
  if (!Number.isFinite(numeric)) return 0
  const factor = 10 ** places
  const scaled = numeric * factor
  const magnitude = Math.abs(scaled)
  const corrected = scaled >= 0
    ? Math.round(magnitude + Number.EPSILON * magnitude)
    : -Math.round(magnitude + Number.EPSILON * magnitude)
  return corrected / factor
}

/**
 * Cost a flat per-unit tariff over a billing period.
 *
 * @param dailyKwh Consumption for ONE day, in kWh.
 * @param tariffPerKwh Flat unit rate in currency per kWh.
 * @param days Length of the billing period in days.
 * @param fixedChargePerPeriod Standing charge added once for the period.
 * @returns `{ ok: true, ... }` with every intermediate quantity, or
 *          `{ ok: false, error }` when an input is not physically sensible.
 */
export function computeBill(dailyKwh, tariffPerKwh, days = DEFAULT_BILLING_DAYS, fixedChargePerPeriod = 0) {
  const daily = Number(dailyKwh)
  const tariff = Number(tariffPerKwh)
  const periodDays = Number(days)
  const fixed = Number(fixedChargePerPeriod) || 0

  if (!Number.isFinite(daily) || daily <= 0) {
    return { ok: false, error: 'Predicted consumption must be greater than zero.' }
  }
  if (!Number.isFinite(tariff) || tariff <= 0) {
    return { ok: false, error: 'Tariff must be greater than zero.' }
  }
  if (!Number.isInteger(periodDays) || periodDays < 1 || periodDays > MAX_BILLING_DAYS) {
    return { ok: false, error: 'Billing period must be a whole number of days between 1 and 365.' }
  }
  if (!Number.isFinite(fixed) || fixed < 0) {
    return { ok: false, error: 'Standing charge cannot be negative.' }
  }

  const consumptionKwh = daily * periodDays
  const energyCharge = consumptionKwh * tariff

  return {
    ok: true,
    periodDays,
    dailyKwh: roundHalfUp(daily, 3),
    consumptionKwh: roundHalfUp(consumptionKwh, 3),
    tariffPerKwh: roundHalfUp(tariff, 4),
    energyCharge: roundHalfUp(energyCharge, 2),
    fixedChargePerPeriod: roundHalfUp(fixed, 2),
    estimatedBill: roundHalfUp(energyCharge + fixed, 2),
  }
}

File: WattWise-AI/src/billing.py
"""Bill arithmetic for WattWise.

This module is the backend authority for how consumption turns into money.
The frontend has a mirror implementation in ``web/src/lib/energy.js``; the two
are kept honest by a shared fixture, ``web/src/lib/billingCases.json``, which
the pytest suite reads and asserts against.

The contract, in one place
--------------------------
For a flat per-unit tariff::

    consumption_kwh = daily_kwh * days
    energy_charge   = consumption_kwh * tariff_per_kwh
    estimated_bill  = energy_charge + fixed_charge_per_period

``fixed_charge_per_period`` is added **once per billing period**, not once
per day. A standing charge billed daily is a different product and is not
modelled here; callers that need it should multiply before calling.

There is no time-of-use banding, no slab/tier structure, no tax and no
standing-charge proration. The tariff is flat, and the UI says so.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal
from typing import Any

DEFAULT_BILLING_DAYS = 30
MAX_BILLING_DAYS = 365
DEFAULT_CURRENCY = "INR"

FORMULA = (
    "consumption_kwh = daily_kwh * days; "
    "energy_charge = consumption_kwh * tariff_per_kwh; "
    "estimated_bill = energy_charge + fixed_charge_per_period"
)


def round_half_up(value: float, places: int = 2) -> float:
    """Round half away from zero, matching JavaScript's ``Math.round``.

    Python's built-in :func:`round` uses banker's rounding, so ``round(590.625, 2)``
    is ``590.62`` while the JavaScript mirror computes ``590.63``. That single
    difference would make the two implementations disagree on every exact
    half-way value, so both sides round half-up explicitly instead.
    """
    quantum = Decimal(1).scaleb(-places)
    return float(Decimal(str(float(value))).quantize(quantum, rounding=ROUND_HALF_UP))



class BillInputError(ValueError):
    """Raised when billing inputs are not physically sensible."""


def compute_bill(
    daily_kwh: float,
    tariff_per_kwh: float,
    days: int = DEFAULT_BILLING_DAYS,
    fixed_charge_per_period: float = 0.0,
    currency: str = DEFAULT_CURRENCY,
) -> dict[str, Any]:
    """Cost a flat-tariff period.

    Parameters
    ----------
    daily_kwh:
        Predicted consumption for **one day**, in kWh.
    tariff_per_kwh:
        Flat unit rate in ``currency`` per kWh.
    days:
        Length of the billing period in days.
    fixed_charge_per_period:
        Any flat standing charge levied once for the whole period.
    """
    daily_kwh = float(daily_kwh)
    tariff_per_kwh = float(tariff_per_kwh)
    fixed_charge_per_period = float(fixed_charge_per_period)

    if daily_kwh <= 0:
        raise BillInputError("predicted_kwh must be greater than zero")
    if tariff_per_kwh <= 0:
        raise BillInputError("tariff_per_kwh must be greater than zero")
    if fixed_charge_per_period < 0:
        raise BillInputError("fixed_charge_per_period cannot be negative")
    if not isinstance(days, int) or days < 1 or days > MAX_BILLING_DAYS:
        raise BillInputError(f"days must be an integer between 1 and {MAX_BILLING_DAYS}")

    consumption_kwh = daily_kwh * days
    energy_charge = consumption_kwh * tariff_per_kwh
    total = energy_charge + fixed_charge_per_period

    return {
        "period_days": days,
        "predicted_daily_kwh": round_half_up(daily_kwh, 3),
        "consumption_kwh": round_half_up(consumption_kwh, 3),
        "tariff_per_kwh": round_half_up(tariff_per_kwh, 4),
        "energy_charge": round_half_up(energy_charge, 2),
        "fixed_charge_per_period": round_half_up(fixed_charge_per_period, 2),
        "estimated_bill": round_half_up(total, 2),
        "currency": currency,
        "tariff_type": "flat",
        "formula": FORMULA,
    }
